models/cow_protocol/ethereum/cow_protocol_ethereum_trades.sql

```diff
@@ -97,57 +97,54 @@ trades_with_token_units as (
 -- This, independent, aggregation defines a mapping of order_uid and trade
--- TODO - create a view for the following block mapping uid to app_data
-order_ids as (
-    select evt_tx_hash, collect_list(orderUid) as order_ids
-    from (  select orderUid, evt_tx_hash, evt_index
-            from {{ source('gnosis_protocol_v2_ethereum', 'GPv2Settlement_evt_Trade') }}
-             {% if is_incremental() %}
-             where evt_block_time >= date_trunc("day", now() - interval '1 week')
-             {% endif %}
-                     sort by evt_index
-         ) as _
-    group by evt_tx_hash
-),
-
-exploded_order_ids as (
-    select evt_tx_hash, posexplode(order_ids)
-    from order_ids
-),
-
-reduced_order_ids as (
+sorted_orders as (
     select
-        col as order_id,
-        -- This is a dirty hack!
-        collect_list(evt_tx_hash)[0] as evt_tx_hash,
-        collect_list(pos)[0] as pos
-    from exploded_order_ids
-    group by order_id
+        evt_tx_hash,
+        evt_block_number,
+        collect_list(orderUid) as order_ids
+    from (
+        select
+            evt_tx_hash,
+            evt_block_number,
+            orderUid
+        from gnosis_protocol_v2_ethereum.GPv2Settlement_evt_Trade
+        {% if is_incremental() %}
+        WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+        {% endif %}
+        distribute by
+            evt_tx_hash, evt_block_number
+        sort by
+            evt_index
+    )
+    group by evt_tx_hash, evt_block_number
 ),
 
-trade_data as (
-    select call_tx_hash,
-           posexplode(trades)
-    from {{ source('gnosis_protocol_v2_ethereum', 'GPv2Settlement_call_settle') }}
-    where call_success = true
-    {% if is_incremental() %}
-    AND call_block_time >= date_trunc("day", now() - interval '1 week')
-    {% endif %}
+orders_and_trades as (
+    select
+        evt_tx_hash,
+        trades,
+        order_ids
+    from sorted_orders
+    inner join {{ source('gnosis_protocol_v2_ethereum', 'GPv2Settlement_call_settle') }}
+        on evt_block_number = call_block_number
+        and evt_tx_hash = call_tx_hash
+-- this is implied by the inner join
+--      and call_success = true
 ),
-
+-- Validate Uid <--> app_data mapping here: https://dune.com/queries/1759039?d=1
 uid_to_app_id as (
     select
-        order_id as uid,
-        get_json_object(trades.col, '$.appData') as app_data,
-        get_json_object(trades.col, '$.receiver') as receiver,
-        get_json_object(trades.col, '$.sellAmount') as limit_sell_amount,
-        get_json_object(trades.col, '$.buyAmount') as limit_buy_amount,
-        get_json_object(trades.col, '$.validTo') as valid_to,
-        get_json_object(trades.col, '$.flags') as flags
-    from reduced_order_ids order_ids
-             join trade_data trades
-                  on evt_tx_hash = call_tx_hash
-                      and order_ids.pos = trades.pos
+        uid,
+        get_json_object(trade, '$.appData') as app_data,
```

Comment:
The primary purpose of this PR is to change the way app_data is constructed. This "fix" here is why I believe the incremental strategy is failing the uniqueness condition.